src/engine.ml

```diff
@@ -906,36 +958,72 @@ let server_send_push_reply config is_not_taken tls session key tls_data =
   in
   Ok (Some (`Established ip_config), config', channel_st, [ (`Control, out) ])
 
-let server_handle_tls_data config is_not_taken session keys tls d =
+let server_handle_tls_data config auth_user_pass is_not_taken session keys tls d
+    =
   let open Result.Syntax in
-  let* next, config, out = kex_server config session keys tls d in
+  let* next, config, out =
+    kex_server config auth_user_pass session keys tls d
+  in
+  let out = List.map (fun out -> (`Control, out)) out in
   match next with
   | `Send_push_reply (tls', key, tls_data) ->
       let* ip_config, config, channel_st, out' =
         server_send_push_reply config is_not_taken tls' session key tls_data
       in
-      Ok (ip_config, config, channel_st, (`Control, out) :: out')
+      Ok (ip_config, config, channel_st, out @ out')
   | `State (channel_st, ip_config) ->
       (* keys established, move forward to "expect push request (reply with push reply)" *)
-      Ok (ip_config, config, channel_st, [ (`Control, out) ])
+      Ok (ip_config, config, channel_st, out)
+  | `Authentication_failed _tls -> Ok (Some `Exit, config, Expect_reset, out)
```

Comment:
Very nice with the `` `Exit `` here. It didn't require as big a change as I thought \o/

I put `Expect_reset` to avoid sort of reset the state. I'm not sure it's sound to reuse the state on a reconnect, and we emit an `` `Exit `` now so the caller shouldn't reuse the state.

Reply:
I agree. Alternatives include:
- allow the error case to carry an out packet
- allow the ok case to have no state (and no config?)

I'd go for the first one if asked, but don't see any advantage right now to refactor.